server/src/main/java/org/elasticsearch/cluster/coordination/ElectionSchedulerFactory.java

```diff
@@ -0,0 +1,203 @@
+        return nonNegative(randomSupplier.getAsLong()) % (upperBound - 1) + 1;
+    }
+
+    @Override
+    public String toString() {
+        return "ElectionSchedulerFactory{" +
+            "initialTimeout=" + initialTimeout +
+            ", backoffTime=" + backoffTime +
+            ", maxTimeout=" + maxTimeout +
+            '}';
+    }
+
+    private class ElectionScheduler implements Releasable {
+        /**
+         * The current maximum timeout: the next election is scheduled randomly no later than this number of milliseconds in the future. On
+         * each election attempt this value is increased by `backoffTime`, up to the `maxTimeout`, to adapt to higher-than-expected latency.
+         */
+        private final AtomicLong currentMaxTimeoutMillis = new AtomicLong(Math.min(initialTimeout.millis(), maxTimeout.millis()));
+        private final AtomicBoolean isClosed = new AtomicBoolean();
+
+        /**
+         * Calculate the next maximum timeout by backing off the current value by `backoffTime` up to the `maxTimeout`.
+         */
+        private long backOffCurrentMaxTimeout(long currentMaxTimeoutMillis) {
+            return Math.min(maxTimeout.getMillis(), currentMaxTimeoutMillis + backoffTime.getMillis());
+        }
+
+        void scheduleNextElection(final TimeValue gracePeriod, final Runnable scheduledRunnable) {
+            if (isClosed.get()) {
+                logger.debug("{} not scheduling election", this);
+                return;
+            }
+
+            final long maxDelayMillis = currentMaxTimeoutMillis.getAndUpdate(this::backOffCurrentMaxTimeout);
+            final long delayMillis = randomPositiveLongLessThan(random::nextLong, maxDelayMillis + 1) + gracePeriod.millis();
+            final Runnable runnable = new AbstractRunnable() {
+                @Override
+                public void onFailure(Exception e) {
+                    logger.debug("unexpected exception in wakeup", e);
+                    assert false : e;
+                }
+
+                @Override
+                protected void doRun() {
+                    if (isClosed.get()) {
+                        logger.debug("{} not starting election", this);
+                        return;
+                    }
+                    logger.debug("{} starting election", this);
+                    scheduledRunnable.run();
+                }
+
+                @Override
+                public void onAfter() {
+                    scheduleNextElection(TimeValue.ZERO, scheduledRunnable);
+                }
+
+                @Override
+                public String toString() {
+                    return "scheduleNextElection{gracePeriod=" + gracePeriod
+                        + ", maxDelayMillis=" + maxDelayMillis
+                        + ", delayMillis=" + delayMillis
+                        + ", " + ElectionScheduler.this + "}";
+                }
+
+                @Override
+                public boolean isForceExecution() {
```

Comment:
@ywelsch I didn't totally understand what action you wanted here (see 
https://github.com/elastic/elasticsearch/pull/32709#discussion_r208713897).